src/com/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -928,119 +757,269 @@
-             WHERE fp.depth = 0 AND
-                   fs.certname = ?"
-     certname]
-    (zipmap (map :name result-set)
-            (map :value result-set))))
-
-(pls/defn-validated insert-facts!
-  "Given a certname and set of fact_value_id's, insert them into the facts table"
-  [factset :- s/Int
-   ids     :- #{s/Int}]
-  (let [default {:factset_id factset}
-        rows    (map #(assoc default :fact_value_id %)
-                     ids)]
-    (apply sql/insert-records :facts rows)))
+  (sql/transaction
+   (let [factset-id (certname-to-factset-id certname)
+         dead-pairs (select-pid-vid-pairs-for-factset factset-id)]
+     (sql/do-commands
+      (format "DELETE FROM facts WHERE factset_id = %s" factset-id))
+     (delete-pending-path-id-orphans! factset-id dead-pairs)
+     (delete-pending-value-id-orphans! factset-id dead-pairs)
+     (sql/delete-rows :factsets ["id=?" factset-id]))))
+
+(defn-validated insert-facts-pv-pairs!
+  [factset-id :- s/Int
+   pairs :- (s/either [(s/pair s/Int "path-id" s/Int "value-id")]
+                      #{(s/pair s/Int "path-id" s/Int "value-id")})]
+  (apply sql/insert-records
+         :facts
+         (for [[pid vid] pairs]
+           {:factset_id factset-id :fact_path_id pid :fact_value_id vid})))
+
+(defn existing-paths->ids
+  "Returns a map from path strings to id for each path that's already
+  in the database."
+  [pathstrs]
+  (into {}
+        (for [{:keys [path id]}
+              (apply query-to-vec
+                     (format "select path, id from fact_paths where path %s"
+                             (jdbc/in-clause pathstrs))
+                     pathstrs)]
+          [path id])))
+
+(defn insert-new-paths!
+  "Inserts the path strings into the database and returns a map from
+  the path strings to their database ids."
+  [pathstrs]
+  (zipmap pathstrs
+          (map :id (apply
+                    sql/insert-records :fact_paths
+                    (map (comp facts/path->pathmap facts/string-to-factpath)
+                         pathstrs)))))
+
+(defn existing-value-hashes->ids
+  "Returns a map from value hash to id for each value that's already
+  in the database."
+  [value-hashes]
+  (into {}
+        (for [{:keys [value_hash id]}
+              (apply query-to-vec
+                     (format "SELECT value_hash, id FROM fact_values
+                                WHERE value_hash %s"
+                             (jdbc/in-clause value-hashes))
+                     value-hashes)]
+          [value_hash id])))
+
+(defn insert-new-values!
+  "Inserts the valuemaps into the database and returns a map from
+  value hash to valuemap.  The returned valuemaps will have an :id."
+  [valuemaps]
+  (let [vhashes (map :value_hash valuemaps)]
+    (zipmap vhashes
+            (let [vmaps valuemaps]
+              (map (fn [id vmap] (assoc vmap :id id))
+                   (map :id (apply sql/insert-records :fact_values valuemaps))
```

Comment:
This seems like a useful chunk of generic functionality on it's own. Given a list of records, insert them and assoc into the record the id.